controlplane/src/core/bufservices/PlatformService.ts

```diff
@@ -3721,6 +3733,120 @@ export default function (opts: RouterOptions): Partial<ServiceImpl<typeof Platfo
+      const logger = opts.logger.child({
+        service: ctx.service.typeName,
+        method: ctx.method.name,
+      });
+
+      return handleError<PlainMessage<PublishPersistedOperationsResponse>>(logger, async () => {
+        const authContext = await opts.authenticator.authenticate(ctx.requestHeader);
+        const userId = authContext.userId;
+        if (!userId) {
+          return {
+            response: {
+              code: EnumStatusCode.ERROR_NOT_AUTHENTICATED,
+              details: `User not found in the authentication context`,
+            },
+            operations: [],
+          };
+        }
+        const organizationId = authContext.organizationId;
+        const federatedGraphRepo = new FederatedGraphRepository(opts.db, organizationId);
+
+        // Validate everything before we update any data
+        const graphSDL = await federatedGraphRepo.getLatestValidSdlOfFederatedGraph(req.graphName);
+        const federatedGraph = await federatedGraphRepo.byName(req.graphName);
+        if (!graphSDL || federatedGraph === undefined) {
+          return {
+            response: {
+              code: EnumStatusCode.ERR_NOT_FOUND,
+              details: `Federated graph '${req.graphName}' does not exist`,
+            },
+            operations: [],
+          };
+        }
+        const graphAST = parse(graphSDL);
+        const graphSchema = buildASTSchema(graphAST);
+        for (const operationContents of req.operations) {
+          let opAST: DocumentNode;
+          try {
+            opAST = parse(operationContents);
+          } catch (e: any) {
+            return {
+              response: {
+                code: EnumStatusCode.ERR,
+                details: `Operation ${operationContents} is not valid: ${e}`,
+              },
+              operations: [],
+            };
+          }
+          const errors = validate(graphSchema, opAST, undefined, { maxErrors: 1 });
+          if (errors.length > 0) {
+            const errorDetails = errors.map((e) => `${e.toString()}`).join(', ');
+            return {
+              response: {
+                code: EnumStatusCode.ERR,
+                details: `Operation "${operationContents}" is not valid: ${errorDetails}`,
+              },
+              operations: [],
+            };
+          }
+        }
+        try {
+          await federatedGraphRepo.registerClient(userId, req.graphName, req.clientName);
+        } catch (e: any) {
+          const message = e instanceof Error ? e.message : e.toString();
+          return {
+            response: {
+              code: EnumStatusCode.ERR,
+              details: `Could not register client "${req.clientName}": ${message}`,
+            },
+            operations: [],
+          };
+        }
+        const operations: PublishedOperation[] = [];
+        for (const operationContents of req.operations) {
+          const operationHash = crypto.createHash('sha256').update(operationContents).digest('hex');
```

Comment:
Why do we need to hash it here again? Can we not just rely on the manifest they provide or hash it client side?

Reply:
If we want to guarantee that the hash matches the contents we can't rely on the hashes reported by the client. Otherwise, operation contents under a given hash might change and we cannot cache them forever.

Reply:
Ok